tsl/src/remote/dist_copy.c

```diff
@@ -254,19 +252,36 @@ static void
 start_remote_copy_on_new_connection(CopyConnectionState *state, TSConnection *connection)
 {
 	PGconn *pg_conn = remote_connection_get_pg_conn(connection);
+
 	if (PQisnonblocking(pg_conn))
 		ereport(ERROR,
 				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
 				 errmsg("distributed copy doesn't support non-blocking connections")));
 
 	if (!list_member_ptr(state->connections_in_use, connection))
 	{
-		PGresult *res = PQexec(pg_conn, state->outgoing_copy_cmd);
+		PGresult *volatile res = NULL;
 
-		if (PQresultStatus(res) != PGRES_COPY_IN)
-			ereport(ERROR,
-					(errcode(ERRCODE_CONNECTION_FAILURE),
-					 errmsg("unable to start remote COPY on data node (%d)", PQresultStatus(res))));
+		PG_TRY();
+		{
+			res = PQexec(pg_conn, state->outgoing_copy_cmd);
+
+			if (PQresultStatus(res) != PGRES_COPY_IN)
```

Comment:
There is a `PQresultErrorMessage(res)` inside the error message that can reference memory inside the `res` structure, so he needs to copy that before clearing it. Otherwise, I agree.

Reply:
A PGresult is allocated by malloc (not PG memory context) and if we do not release it will cause a permanent memory leak. We need to guarantee that the PGresult is released even in case of an unexpected error (perhaps thrown deep in a function called during the lifetime of the PGresult.

The only way to guarantee that the PGresult is released in case of errors is to allocate within a PG_TRY context, and release it in the error handler. It also solves the problem @mkindahl mentions that we need to reference the error data in the error we are throwing (otherwise we need to make copies of all the error data first).

The only other alternative is to copy all data from the PGresult.